Custom model and task feature

anishabhatnagar opened the pull request:
# Added model & task selection, callbacks, and state handling

## Summary  
Introduces UI controls for selecting models and tasks, wires up callbacks to reload data and compute embeddings, and adds state components to track texts for later highlighting. Embedding generation for the default LUAR‑MUD ([rrivera1849/LUAR-MUD](https://huggingface.co/rrivera1849/LUAR-MUD)) model is currently failing; therefore, graph displays are still showing the HRS defaults, please ignore them for now.

## Changes  
1. **Model selection radio** (lines 104–125 in `app.py`)  
   - Choices:  
     - `rrivera1849/LUAR-MUD` (default)  
     - **Other** → shows a text box for any Hugging Face Model ID  
2. **Task source radio** (lines 127–178)  
   - **Predefined HRS Task** → choose from dropdown  
   - **Upload Your Own Task** → upload `.txt` files (testing with Federalist dataset)  
[candidate_1_Jay.txt](https://github.com/user-attachments/files/21282872/candidate_1_Jay.txt)
[candidate_2_Madison.txt](https://github.com/user-attachments/files/21282873/candidate_2_Madison.txt)
[candidate_3_hamilton.txt](https://github.com/user-attachments/files/21282874/candidate_3_hamilton.txt)
[mystery_text_hamilton.txt](https://github.com/user-attachments/files/21282876/mystery_text_hamilton.txt)

3. **Callbacks**  
   - Reload task display & compute embeddings when:  
     - Model changes (radio or custom‑ID textbox)  
     - Task choice changes (dropdown or file upload)  
   - `update_task_display()` , `toggle_task()`,` toggle_custom_model()` in `utils/ui.py `
4. **Gradio State components**  
   - Track raw texts uploaded by the user for end‑of‑flow highlighting  
5. **Embedding issue documented**  
   - Attempting embeddings with “rrivera1849/LUAR‑MUD” fails:  
     ```
     'LUARConfig' object has no attribute 'hidden_size'
     ```  
6. **Temporary graph behavior**  
   - Until embeddings are fixed, the plot still shows HRS default mystery/candidate authors (task 0). Please ignore those markers for now.

## TBD
- @MiladAlshomary please take a look at the embedding generation code, i dont know if i introduced a bug or I am looking at the wrong code.  
- Candidate ids seem to be hardcoded in the span highlight section
- guide strings/ explantory text throught out the demo needs to be updated.

MiladAlshomary commented:
I have pushed some updates. The main issue with the code that you were using was the LUAR model, which is not a sentence transformer. Instead, any model to be used should be a sentence transformer. For example, gabrielloiseau/LUAR-MUD-sentence-transformers, which is a sentence transformer version of LUAR.

However, I also noticed some other issues that I fixed. You were embedding only the texts of the task users not the full background author dataframe. Now, I modified the code so I add the task authors into the background_authors_dataframe and run the embedding on all. But I still didn't fully test the functionality.

anishabhatnagar commented:
@MiladAlshomary 
I have added -
1. `gradio.state()` blocks to hold the embeddings for Task authors (`task_authors_embeddings_df`) and background authors (`background_authors_embeddings_df`)
2. added caching to ensure the same set of text is not processed twice by the same model: wrapper around `generate_style_embedding()` -> `cached_generate_style_embedding()`
3. Added a "Load Task and Generate Embeddings" button, to avoid triggering the on change handlers for the model/task definition with incomplete inputs. Once the user selects the model and task, must click the button to generate the embeddings. 
4. updated the visualization code to plot the new embeddings (@MiladAlshomary please check my updated logic once .. thanks!)(Have left the old code in the comments for now, will clean up with the rest of the refactoring)
5.  changed a hardcoded value in the highlights section : "Candidate 1" was **always** tagged as "Predicted Author", regardless of it being one or not. Which leads to the question -> how to get the predicted author for custom texts/ new models ? 
6. Also updated the guide/instructional strings in a few places.
